drivers/rail/rail_drv.c

```diff
@@ -0,0 +1,798 @@
+    if (bRet != true) {
+        DEBUG("Can not set short addr\n");
+    }
+
+    // set long addr
+    DEBUG("Set LongAddr 0x%08lx%08lx\n", ntohl(dev->eui.uint64.u32[0]), ntohl(dev->eui.uint64.u32[1]));
+
+    memcpy(netdev->long_addr, &dev->eui.uint8, IEEE802154_LONG_ADDRESS_LEN);
+
+    // reversed order, because thats what rails want
+    uint64_t addr_rev = byteorder_ntohll(dev->eui.uint64);
+
+    bRet = RAIL_IEEE802154_SetLongAddress(dev->rhandle, (uint8_t *) &addr_rev, 0);
+    if (bRet != true) {
+        DEBUG("Can not set long addr\n");
+    }
+
+
+    // get transmitt power
+    RAIL_TxPowerLevel_t power_level_tx = RAIL_GetTxPower(dev->rhandle);
+    RAIL_TxPower_t power_tx_ddBm = RAIL_ConvertRawToDbm(dev->rhandle,
+    #if RAIL_RADIO_BAND == 2400
+                                                        RAIL_TX_POWER_MODE_2P4_HP, // 2.4GHZ HighPower, TODO low power?
+    #elif (RAIL_RADIO_BAND == 868) || (RAIL_RADIO_BAND == 915)
+                                                        RAIL_TX_POWER_MODE_SUBGIG,
+    #endif
+                                                        power_level_tx
+                                                        );
+
+
+
+    DEBUG("TX Power set to raw: %hhu %hd deci dBm\n", power_level_tx, power_tx_ddBm);
+
+
+    // config events
+    // it is possible to get all events with RAIL_EVENTS_ALL
+
+    ret = RAIL_ConfigEvents(dev->rhandle,
+                            RAIL_EVENTS_ALL,    // mask of events should be mod
+                                                // events to subsripe
+                            RAIL_EVENT_RX_ACK_TIMEOUT   |
+                            RAIL_EVENT_RX_FRAME_ERROR   |
+                            RAIL_EVENT_RX_ADDRESS_FILTERED |
+                            RAIL_EVENT_RX_PACKET_RECEIVED |
+                            RAIL_EVENT_TX_PACKET_SENT   |
+                            RAIL_EVENT_TX_CHANNEL_BUSY  |
+                            RAIL_EVENT_TX_ABORTED       |
+                            RAIL_EVENT_TX_BLOCKED       |
+                            RAIL_EVENT_TX_UNDERFLOW     |
+                            RAIL_EVENT_IEEE802154_DATA_REQUEST_COMMAND |
+                            RAIL_EVENT_CAL_NEEDED
+                            );
+
+    if (ret != RAIL_STATUS_NO_ERROR) {
+        LOG_ERROR("Can not subsripe to rail events - error msg: %s", rail_error2str(ret));
+        return -1;
+    }
+
+    // put radio to idle mode
+    // TODO not needed anymore?
+    // RAIL_Idle(dev->rhandle, RAIL_IDLE, true);
+
+    // dev->state = RAIL_TRANSCEIVER_STATE_IDLE;
+
+
+
+    DEBUG("rail_init done\n");
+    LOG_INFO("rail radio initialised\n");
+
+    return 0;
+}
+
+int rail_tx_prepare(rail_t *dev)
+{
+    // check state
+    // if uninit/init etc/ returnerror;
+    // if calibrate, error
+    // tx error
+    // if idle and rx
+    //    if waiting for ack
+    //      no timeout? error
+    //      timeout?, stats++, cont
+
+    dev->state = RAIL_TRANSCEIVER_STATE_TX;
```

Comment:
well, I suppose the commented logic is missing? If you have doubts check again the at86rf2xx driver.

Reply:
I probably doesn't need this function at all (it is there, because I started with the at86rf2xx driver ;) ). Or at least atm.